File: core/steam_paths.py
"""Locate the Steam installation and all of its library folders.

Cross-platform: Windows reads the registry, Linux/macOS check the usual dirs.
Library folders are then read from libraryfolders.vdf (a single Steam install
can spread games across several drives).
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

from . import vdf


def _steam_root_windows() -> Path | None:
    try:
        import winreg
    except ImportError:
        return None
    candidates = [
        (winreg.HKEY_CURRENT_USER, r"Software\Valve\Steam", "SteamPath"),
        (winreg.HKEY_LOCAL_MACHINE, r"SOFTWARE\WOW6432Node\Valve\Steam", "InstallPath"),
        (winreg.HKEY_LOCAL_MACHINE, r"SOFTWARE\Valve\Steam", "InstallPath"),
    ]
    for hive, subkey, value in candidates:
        try:
            with winreg.OpenKey(hive, subkey) as key:
                path, _ = winreg.QueryValueEx(key, value)
                p = Path(path)
                if p.exists():
                    return p
        except OSError:
            continue
    return None


def _steam_root_unix() -> Path | None:
    home = Path.home()
    candidates = [
        home / ".steam" / "steam",
        home / ".steam" / "root",
        home / ".local" / "share" / "Steam",
        home / "Library" / "Application Support" / "Steam",  # macOS
    ]
    for p in candidates:
        if (p / "steamapps").exists() or (p / "config").exists():
            return p.resolve()
    return None


def steam_root() -> Path | None:
    """Return the root Steam install directory, or None if not found."""
    env = os.environ.get("STEAM_ROOT")
    if env and Path(env).exists():
        return Path(env)
    if sys.platform.startswith("win"):
        return _steam_root_windows()
    return _steam_root_unix()


def steam_executable(root: Path | None = None) -> Path | None:
    """Path to the Steam executable (used to start Steam and to run -shutdown)."""
    root = root or steam_root()
    if not root:
        return None
    if sys.platform.startswith("win"):
        exe = root / "steam.exe"
        return exe if exe.exists() else None
    # Linux/macOS usually have `steam` on PATH; fall back to that.
    local = root / "steam.sh"
    if local.exists():
        return local
    return Path("steam")


def library_folders(root: Path | None = None) -> list[Path]:
    """All Steam library folders (each contains a steamapps/ dir)."""
    root = root or steam_root()
    if not root:
        return []
    libs: list[Path] = []
    for candidate in (
        root / "steamapps" / "libraryfolders.vdf",
        root / "config" / "libraryfolders.vdf",
    ):
        if candidate.exists():
            data = vdf.load(candidate)
            folders = data.get("libraryfolders", data)
            for _, entry in folders.items():
                if isinstance(entry, dict) and "path" in entry:
                    p = Path(entry["path"])
                    if p.exists() and p not in libs:
                        libs.append(p)
            break
    # Always include the root install as a library.
    if root not in libs and (root / "steamapps").exists():
        libs.insert(0, root)
    return libs


if __name__ == "__main__":
    r = steam_root()
    print("Steam root:", r)
    print("Executable:", steam_executable(r))
    print("Libraries:")
    for lib in library_folders(r):
        print("  -", lib)

File: src/core/steam_paths.cpp
#include "steam_paths.h"

#include "platform.h"
#include "vdf.h"

#include <algorithm>
#include <fstream>
#include <sstream>

namespace ss::steam {

namespace {

#if defined(_WIN32)
std::optional<fs::path> steamRootWindows() {
    using platform::Hive;
    struct Cand { Hive hive; const char* subkey; const char* value; };
    const Cand cands[] = {
        {Hive::CurrentUser, "Software\\Valve\\Steam", "SteamPath"},
        {Hive::LocalMachine, "SOFTWARE\\WOW6432Node\\Valve\\Steam", "InstallPath"},
        {Hive::LocalMachine, "SOFTWARE\\Valve\\Steam", "InstallPath"},
    };
    for (const auto& c : cands) {
        if (auto v = platform::regReadString(c.hive, c.subkey, c.value)) {
            fs::path p(*v);
            if (fs::exists(p)) return p;
        }
    }
    return std::nullopt;
}
#else
std::optional<fs::path> steamRootUnix() {
    fs::path home(platform::homeDir());
    const fs::path cands[] = {
        home / ".steam" / "steam",
        home / ".steam" / "root",
        home / ".local" / "share" / "Steam",
        home / "Library" / "Application Support" / "Steam",  // macOS
    };
    for (const auto& p : cands) {
        std::error_code ec;
        if (fs::exists(p / "steamapps", ec) || fs::exists(p / "config", ec))
            return fs::canonical(p, ec);
    }
    return std::nullopt;
}
#endif

}  // namespace

std::optional<fs::path> steamRoot() {
    if (auto env = platform::getEnv("STEAM_ROOT")) {
        fs::path p(*env);
        if (fs::exists(p)) return p;
    }
#if defined(_WIN32)
    return steamRootWindows();
#else
    return steamRootUnix();
#endif
}

std::optional<fs::path> steamExecutable(const std::optional<fs::path>& rootIn) {
    auto root = rootIn ? rootIn : steamRoot();
    if (!root) return std::nullopt;
#if defined(_WIN32)
    fs::path exe = *root / "steam.exe";
    if (fs::exists(exe)) return exe;
    return std::nullopt;
#else
    fs::path sh = *root / "steam.sh";
    if (fs::exists(sh)) return sh;
    return fs::path("steam");
#endif
}

std::vector<fs::path> libraryFolders(const std::optional<fs::path>& rootIn) {
    auto root = rootIn ? rootIn : steamRoot();
    std::vector<fs::path> libs;
    if (!root) return libs;

    const fs::path candidates[] = {
        *root / "steamapps" / "libraryfolders.vdf",
        *root / "config" / "libraryfolders.vdf",
    };
    for (const auto& cand : candidates) {
        if (!fs::exists(cand)) continue;
        vdf::Value data = vdf::load(cand.string());
        const vdf::Value* folders = data.get("libraryfolders");
        if (!folders) folders = &data;
        for (const auto& kv : folders->map) {
            const vdf::Value& entry = kv.second;
            if (!entry.is_map) continue;
            std::string p = entry.getStr("path");
            if (p.empty()) continue;
            fs::path lp(p);
            std::error_code ec;
            if (fs::exists(lp, ec) &&
                std::find(libs.begin(), libs.end(), lp) == libs.end())
                libs.push_back(lp);
        }
        break;
    }
    // Always include the root install as a library.
    std::error_code ec;
    if (std::find(libs.begin(), libs.end(), *root) == libs.end() &&
        fs::exists(*root / "steamapps", ec))
        libs.insert(libs.begin(), *root);
    return libs;
}

}  // namespace ss::steam
